File: prism-kotlin-sdk/credentials/src/commonMain/kotlin/io/iohk/atala/prism/kotlin/credentials/content/CredentialContent.kt
package io.iohk.atala.prism.kotlin.credentials.content

import io.iohk.atala.prism.kotlin.identity.DID
import kotlinx.serialization.json.*
import kotlin.js.JsExport

/**
 * Credential content JSON representation,
 * this class allows to access nested fields
 * with 'dot' notation like: credentialSubject.fieldName
 *
 * Example:
 * {{{
 *   import kotlinx.serialization.json.*
 *   val credentialContent = buildCredentialContent {
 *     put("issuerDid", "did")
 *     putJsonObject("credentialSubject") {
 *       put("fieldName", 123)
 *     }
 *   }
 *   credentialContent.getString("issuerDid") == "did"
 *   credentialContent.getInt("credentialSubject.fieldName") == 123
 * }}}
 */
@JsExport
data class CredentialContent(val fields: JsonObject) {

    fun getString(field: String): String? =
        getField(field)?.jsonPrimitive?.contentOrNull

    fun getInt(field: String): Int? =
        getField(field)?.jsonPrimitive?.int

    fun getBoolean(field: String): Boolean? =
        getField(field)?.jsonPrimitive?.boolean

    fun getArray(field: String): JsonArray? =
        getField(field)?.jsonArray

    /**
     * Get nested json primitives with 'dot' notation like: credentialSubject.fieldName
     */
    fun getField(fieldName: String): JsonElement? {
        val path = fieldName.split('.')
        return path.drop(1).fold(fields[path[0]]) { field, key ->
            field?.jsonObject?.get(key)
        }
    }

    // Predefined fields
    fun getIssuerDid(): DID? =
        getString(JsonField.IssuerDid.value)?.let {
            DID.fromString(it)
        }

    fun getIssuanceKeyId(): String? =
        getString(JsonField.IssuanceKeyId.value)

    fun getCredentialSubject(): String? =
        getString(JsonField.CredentialSubject.value)

    companion object {
        fun fromString(value: String): CredentialContent =
            CredentialContent(Json.parseToJsonElement(value).jsonObject)
    }
}

enum class JsonField(val value: String) {
    CredentialType("type"),
    Issuer("issuer"),
    IssuerDid("id"),
    IssuerName("name"),
    IssuanceKeyId("keyId"),
    IssuanceDate("issuanceDate"),
    ExpiryDate("expiryDate"),
    CredentialSubject("credentialSubject")
}

/**
 * Create a new CredentialContent instance using buildJsonObject syntax.
 */
inline fun buildCredentialContent(builderAction: JsonObjectBuilder.() -> Unit) =
    CredentialContent(buildJsonObject(builderAction))

File: prism-sdk/credentials/shared/src/main/scala/io/iohk/atala/prism/credentials/content/CredentialContent.scala
package io.iohk.atala.prism.credentials.content

import io.iohk.atala.prism.credentials.content.CredentialContent.JsonFields
import io.iohk.atala.prism.identity.DID

import io.iohk.atala.prism.credentials.content.CredentialContent._

case class CredentialContent(fields: Fields) {

  /**
    * Access credential field by name.
    */
  def apply(field: String): Either[CredentialContentException, Value] = getValue(field)

  /**
    * Access credential field by name.
    */
  def getValue(field: String): Either[CredentialContentException, Value] = {
    val path = field.split('.')
    path.drop(1).foldLeft(fields.find(_.name == path(0)).map(_.value)) {
      case (Some(sf: SubFields), key) => sf.value.find(_.name == key).map(_.value)
      case _ => None
    } match {
      case Some(value) => Right(value)
      case None => Left(FieldNotFoundException(s"Field not found: $field"))
    }
  }

  def getString(field: String): Either[CredentialContentException, String] = {
    getValue(field).flatMap {
      case StringValue(value) => Right(value)
      case _ => Left(WrongTypeException(s"$field is not a String."))
    }
  }

  def getInt(field: String): Either[CredentialContentException, Int] = {
    getValue(field).flatMap {
      case IntValue(value) => Right(value)
      case _ => Left(WrongTypeException(s"$field is not an Int."))
    }
  }

  def getBoolean(field: String): Either[CredentialContentException, Boolean] = {
    getValue(field).flatMap {
      case BooleanValue(value) => Right(value)
      case _ => Left(WrongTypeException(s"$field is not an Boolean."))
    }
  }

  def getSeq(field: String): Either[CredentialContentException, Seq[Any]] = {
    getValue(field).flatMap {
      case SeqValue(values) => Right(values.map(getValue))
      case _ => Left(WrongTypeException(s"$field is not an Int."))
    }
  }

  def getSubFields(field: String): Either[CredentialContentException, IndexedSeq[(String, Any)]] = {
    getValue(field).flatMap {
      case SubFields(fields) => Right(fields.map(f => (f.name, getValue(f.value))))
      case _ => Left(WrongTypeException(s"$field is not an Int."))
    }
  }

  // Predefined fields
  val issuerDid: Either[CredentialContentException, DID] =
    getString(JsonFields.IssuerDid.field).flatMap(did =>
      DID.fromString(did).toRight(WrongTypeException("IssuerDid is not a string."))
    )
  val issuanceKeyId: Either[CredentialContentException, String] = getString(JsonFields.IssuanceKeyId.field)
  val credentialSubject: Either[CredentialContentException, String] = getString(JsonFields.CredentialSubject.field)

  /**
    * Get the underlying value from [[Value]] class.
    */
  private def getValue(value: Value): Any = {
    value match {
      case StringValue(value) => value
      case IntValue(value) => value
      case BooleanValue(value) => value
      case SeqValue(values) => values.map(getValue)
      case SubFields(fields) => fields.map(f => (f.name, getValue(f.value)))
    }
  }
}
object CredentialContent {
  def apply(fields: Field*): CredentialContent = CredentialContent(IndexedSeq(fields: _*))
  def empty: CredentialContent = CredentialContent(IndexedSeq.empty)

  sealed trait Value extends Any { def value: Any }
  case class StringValue(value: String) extends AnyVal with Value
  case class IntValue(value: Int) extends AnyVal with Value
  case class BooleanValue(value: Boolean) extends AnyVal with Value
  case class SeqValue(value: Values) extends AnyVal with Value
  case class SubFields(value: Fields) extends AnyVal with Value

  case class Field(name: String, value: Value)

  type Values = Seq[Value]
  def Values(xs: Value*): Values = Seq(xs: _*)

  type Fields = IndexedSeq[Field]
  def Fields(xs: Field*): Fields = IndexedSeq(xs: _*)

  sealed class CredentialContentException(message: String) extends Exception(message)
  case class FieldNotFoundException(message: String) extends CredentialContentException(message)
  case class WrongTypeException(message: String) extends CredentialContentException(message)

  object JsonFields {
    sealed abstract class Field(val field: String)
    case object CredentialType extends Field("type")
    case object Issuer extends Field("issuer")
    case object IssuerDid extends Field("id")
    case object IssuerName extends Field("name")
    case object IssuanceKeyId extends Field("keyId")
    case object IssuanceDate extends Field("issuanceDate")
    case object ExpiryDate extends Field("expiryDate")
    case object CredentialSubject extends Field("credentialSubject")
  }
}
